### ISSUE TITLE
[3.27.0][Android][Impeller] Crash on `fml_check` at `surface_texture_external_texture.cc:55`

### ISSUE THREAD

**Issue by bc-lee** [flutter/flutter#160409]
### Steps to reproduce

- Use a Samsung Galaxy S10 (SM-G973N) / Galaxy S10 5G (SM-G977N) running Android 12 (Build Number: SP1A.210812.016).
- Detailed reproduction steps are pending (time needed to reduce the issue for public sharing).

### Expected results

The external texture from the camera should display correctly.

### Actual results

The app crashes on `fml_check` at `surface_texture_external_texture.cc:55`. Detailed logs are provided below.

### Code sample

To be shared after further reduction.

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[ERROR:flutter/shell/platform/android/surface_texture_external_texture_vk_impeller.cc(122)] Break on 'ImpellerValidationBreak' to inspect point of failure: Invalid external texture.
[FATAL:flutter/shell/platform/android/surface_texture_external_texture.cc(55)] Check failed: state_ == AttachmentState::kAttached.
Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 3436 (1.raster), pid 3243 (foo.bar.baz)
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
Build fingerprint: 'samsung/beyond1lteks/beyond1:12/SP1A.210812.016/G973NKSU7HVG5:user/release-keys'
Revision: '26'
ABI: 'arm64'
Processor: '5'
Timestamp: 2024-12-17 14:04:27.261733329+0900
Process uptime: 15s
Cmdline: foo.bar.baz
pid: 3243, tid: 3436, name: 1.raster  >>> foo.bar.baz <<<
uid: 10451
signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
Abort message: '[FATAL:flutter/shell/platform/android/surface_texture_external_texture.cc(55)] Check failed: state_ == AttachmentState::kAttached.
'
    x0  0000000000000000  x1  0000000000000d6c  x2  0000000000000006  x3  0000007a03216fa0
    x4  0000007d954ce000  x5  0000007d954ce000  x6  0000007d954ce000  x7  0000000000522748
    x8  00000000000000f0  x9  0000007d91f53280  x10 ffffff00fffffbdf  x11 0000000000000001
    x12 0000fffffffff3ff  x13 0000000000000085  x14 0000000000000000  x15 00004c00949ca05e
    x16 0000007d91ff4058  x17 0000007d91fd0da0  x18 00000079f9980000  x19 00000000000000ac
    x20 0000000000000cab  x21 00000000000000b2  x22 0000000000000d6c  x23 00000000ffffffff
    x24 0000007a032176a0  x25 0000007a032176e0  x26 0000007a03217580  x27 0000000000000059
    x28 0000000000000001  x29 0000007a03217020
    lr  0000007d91f8233c  sp  0000007a03216f80  pc  0000007d91f8236c  pst 0000000000000000
backtrace:
      #00 pc 000000000005136c  /apex/com.android.runtime/lib64/bionic/libc.so (abort+180) (BuildId: eb0c8b7f827292af83855e89b431276a)
      #01 pc 0000000001c82a34  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #02 pc 0000000001cae4ac  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #03 pc 0000000001c9307c  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #04 pc 0000000001fd5c38  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #05 pc 0000000001fd063c  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #06 pc 0000000001fd5f60  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #07 pc 0000000001fd063c  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #08 pc 0000000001fd5f60  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #09 pc 0000000001fd063c  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #10 pc 0000000001fd5f60  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #11 pc 0000000001fd063c  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #12 pc 0000000001fd5f60  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #13 pc 0000000001fd063c  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #14 pc 0000000001fd5f60  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #15 pc 0000000001fd063c  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #16 pc 0000000001fd4654  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #17 pc 000000000211e824  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #18 pc 0000000001caf0a8  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #19 pc 0000000001cb4b44  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #20 pc 0000000000016bb0  /system/lib64/libutils.so (android::Looper::pollInner(int)+912) (BuildId: 176a88c64d70f517a4c1c58b40932d90)
      #21 pc 00000000000167b8  /system/lib64/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+112) (BuildId: 176a88c64d70f517a4c1c58b40932d90)
      #22 pc 0000000000017284  /system/lib64/libandroid.so (ALooper_pollOnce+100) (BuildId: 0e0a06d97c3e68482b682ccc78aaf297)
      #23 pc 0000000001cb4acc  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #24 pc 0000000001caeff4  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #25 pc 0000000001cb2db0  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #26 pc 0000000001cb2c48  /data/app/~~gt6Pxms1R7pIqyakxBJc-w==/foo.bar.baz/base.apk!libflutter.so (BuildId: 4619b3c8fc7f4409fb7b46ab3519f15bad2c2792)
      #27 pc 00000000000b4ad8  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+264) (BuildId: eb0c8b7f827292af83855e89b431276a)
      #28 pc 0000000000052c08  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: eb0c8b7f827292af83855e89b431276a)
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[!] Flutter (Channel [user-branch], 3.27.0, on macOS 14.7.1 23H222 darwin-arm64,
    locale en-KR)
    ! Flutter version 3.27.0 on channel [user-branch] at /Users/redacted/flutter
      Currently on an unknown channel. Run `flutter channel` to switch to an
      official channel.
      If that doesn't fix the issue, reinstall Flutter by following instructions
      at https://flutter.dev/setup.
    ! Upstream repository unknown source is not a standard remote.
      Set environment variable "FLUTTER_GIT_URL" to unknown source to dismiss
      this error.
    • Framework revision 8495dee1fd (6 days ago), 2024-12-10 14:23:39 -0800
    • Engine revision 83bacfc525
    • Dart version 3.6.0
    • DevTools version 2.40.2
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/redacted/android/sdk
    • Platform android-35, build-tools 35.0.0
    • ANDROID_HOME = /Users/redacted/android/sdk
    • Java binary at: /Users/redacted/Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 21.0.5+-12771278-b631.28)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16C5032a
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version unknown)
    • Android Studio at /Users/redacted/Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to determine Android Studio version.
    • android-studio-dir = /Users/redacted/Applications/Android Studio.app
    • Java version OpenJDK Runtime Environment (build 21.0.5+-12771278-b631.28)

[✓] IntelliJ IDEA Ultimate Edition (version 2024.3.1)
    • IntelliJ at /Users/redacted/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 83.0.4
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version unknown)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.102.0
    ✗ Unable to determine VS Code version.

[✓] Connected device (4 available)
    • SM G973N (mobile)               • R39M601C0TV           • android-arm64  •
      Android 12 (API 31)
    • macOS (desktop)                 • macos                 • darwin-arm64   •
      macOS 14.7.1 23H222 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         •
      macOS 14.7.1 23H222 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript •
      Google Chrome 131.0.6778.140

[✓] Network resources
    • All expected network resources are available.
```

</details>

### Additional Context

Basically, this report is same with https://github.com/flutter/flutter/issues/159324 , except that this Crash happens on Release channel.
I also checked that `3.28.0-0.1.pre` (3e493a3e4d0a5c99fa7da51faae354e95a9a1abe) don't have this issue.
Please Flutter team members open CP request to fix this issue on stable channel by using merged PR: https://github.com/flutter/engine/commit/f852b36b22ebf2b89473fbc0c91b7c88bec09f3e

**Comment by bc-lee** [comment]
Progress is being made on the PR: flutter/engine#57249. I'll keep this open for now.

**Comment by Guo8a** [comment]
This issue been resolved in version 3.27.1?

**Comment by bc-lee** [comment]
> This issue been resolved in version 3.27.1?

No. While the patch (https://github.com/flutter/engine/pull/57249) has been submitted, the corresponding CP request (https://github.com/flutter/flutter/issues/160460) was not approved by the release team.

**Comment by Guo8a** [comment]
> > This issue been resolved in version 3.27.1?
> 
> No. While the patch ([flutter/engine#57249](https://github.com/flutter/engine/pull/57249)) has been submitted, the corresponding CP request (#160460) was not approved by the release team.

OK, well. If the Flutter version doesn’t get updated, how do you resolve this issue?

**Comment by bc-lee** [comment]
The workaround is to disable Impeller, which can help avoid rendering issues like this. Refer to [the official Flutter documentation](https://docs.flutter.dev/perf/impeller#android) for instructions.

**Comment by Guo8a** [comment]
> The workaround is to disable Impeller, which can help avoid rendering issues like this. Refer to [the official Flutter documentation](https://docs.flutter.dev/perf/impeller#android) for instructions.

ok, thx~

**Comment by corepuncher** [comment]
I disabled impeller but my app is still crashing in 3.27.1.   I wonder if there is something else going on.  The OOM errors definitely happen for me frequently with impeller enabled on Android.  And I am using the very latest of everything on a new device.

**Comment by bc-lee** [comment]
This issue is not related to OOM errors. If you're encountering frequent OOM issues, could you please file a separate issue with relevant details?

**Comment by corepuncher** [comment]
> This issue is not related to OOM errors. If you're encountering frequent OOM issues, could you please file a separate issue with relevant details?

Thanks. I filed an issue on the OOM errors #160941 .
